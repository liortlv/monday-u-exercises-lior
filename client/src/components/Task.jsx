import taskCSS from "./Task.module.css";
import Button from "./Button";
import Checkbox from "./Checkbox";
import { useCallback, useMemo } from "react";
import api from "../clients/item-client.js";
import { motion } from "framer-motion";
import PropTypes from "prop-types";

function Task({ task, setTasks, tasks }) {
   const addPokemonImage = useMemo(() => {
      return <img src={task.imageURL} className={taskCSS["pokemon-image"]} alt="pokemon" />;
   }, [task.imageURL]);

   const deleteButtonHandler = useCallback(async () => {
      try {
         await api.deleteTask(task.id);
         setTasks(tasks.filter((item) => item.id !== task.id));
      } catch (error) {
         console.error(error);
      }
   }, [setTasks, task.id, tasks]);

   return (
      <motion.li
         data-id={task.id}
         className={taskCSS["task-container"]}
         whileHover={{ scale: 1.05, transition: { duration: 0.15 } }}
         initial={{ x: -130, opacity: 0.3, scale: 0.6 }}
         animate={{ x: 0, opacity: 1, scale: 1 }}
      >
         <Checkbox task={task} tasks={tasks} setTasks={setTasks} />
         <div className={taskCSS.task}>
            {task.text}
            {task.imageURL ? addPokemonImage : ""}
         </div>
         <Button
            onClick={deleteButtonHandler}
            innerText={<i className={`${taskCSS["fa-trash"]} fa-trash fa-solid`}></i>}
            className={taskCSS.delete}
         />
      </motion.li>
   );
}

Task.propTypes = {
   task: PropTypes.object,
   tasks: PropTypes.array,
   setTasks: PropTypes.func,
};

export default Task;
