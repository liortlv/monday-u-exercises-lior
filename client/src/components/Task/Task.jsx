import taskCSS from "./Task.module.css";
import Button from "../Button";
import CheckboxConnector from "../Checkbox/CheckboxConnector";
import { useCallback, useMemo } from "react";
import { motion } from "framer-motion";
import PropTypes from "prop-types";
import { toast } from "react-toastify";
import toastOptions from "../../clients/toast-options";

function Task({ task, deleteTaskAction, undoDeleteTaskAction, getAPITasksAction, setTasksAction }) {
   const addPokemonImage = useMemo(() => {
      return <img src={task.imageURL} className={taskCSS["pokemon-image"]} alt="pokemon" />;
   }, [task.imageURL]);

   const deleteButtonHandler = useCallback(async () => {
      try {
         await deleteTaskAction(task.id);
         toast.error(<Undo />, toastOptions);
      } catch (error) {
         console.error(error);
      }
   }, [task.id, deleteTaskAction]);

   const Undo = () => {
      const handleClick = async () => {
         await undoDeleteTaskAction();
         const response = await getAPITasksAction();
         const allTasks = response.payload;
         allTasks ? setTasksAction(allTasks) : console.error("Server error");
      };

      return (
         <div className={taskCSS["deleted-message"]}>
            Row Deleted{" "}
            <button onClick={handleClick} className={taskCSS["undo-btn"]}>
               Undo
            </button>
         </div>
      );
   };

   return (
      <motion.li
         data-id={task.id}
         className={taskCSS["task-container"]}
         whileHover={{ scale: 1.05, transition: { duration: 0.15 } }}
         initial={{ x: -130, opacity: 0.3, scale: 0.6 }}
         animate={{ x: 0, opacity: 1, scale: 1 }}
      >
         <CheckboxConnector task={task} />
         <div className={taskCSS.task}>
            {task.text}
            {task.imageURL ? addPokemonImage : ""}
         </div>
         <Button
            onClick={deleteButtonHandler}
            innerText={<i className={`${taskCSS["fa-trash"]} fa-trash fa-solid`}></i>}
            className={taskCSS.delete}
         />
      </motion.li>
   );
}

Task.propTypes = {
   task: PropTypes.object,
   deleteTaskAction: PropTypes.func,
   undoDeleteTaskAction: PropTypes.func,
};

export default Task;
