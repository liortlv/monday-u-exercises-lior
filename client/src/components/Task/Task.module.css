.task-container {
   display: flex;
   align-items: center;
}

.task-container:hover > .task {
   cursor: pointer;
   background-color: rgb(0, 200, 255);
}

.task-container:hover > .delete {
   visibility: visible;
}

.task {
   width: 43rem;
   height: 3.5rem;
   background-color: rgb(139, 218, 255);
   font-size: 2rem;
   padding: 1rem;
   margin: 0.3rem;
   display: flex;
   align-items: center;
   justify-content: space-between;
   border-radius: 3px;
}

.delete {
   background-color: red;
   width: 3.5rem;
   height: 3.5rem;
   border-radius: 3px;
   display: flex;
   align-items: center;
   justify-content: center;
   margin-top: 0.3rem;
   visibility: hidden;
   border: none;
}

.delete:hover {
   cursor: pointer;
   background-color: rgb(221, 0, 0);
}

.fa-trash {
   font-size: 1.6rem;
}

.pokemon-image {
   width: 4.5rem;
}
