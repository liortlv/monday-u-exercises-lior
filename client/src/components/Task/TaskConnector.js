import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import Task from "./Task";
import { deleteTaskAction } from "../../actions/api-actions";

const mapDispatchToProps = (dispatch, ownProps) => {
   return bindActionCreators({ deleteTaskAction }, dispatch);
};

export default connect(mapDispatchToProps)(Task);
