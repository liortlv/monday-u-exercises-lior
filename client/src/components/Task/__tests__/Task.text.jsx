import Task from "../Task";
import ShallowRenderer from "react-test-renderer/shallow";

it("renders pokemon task correctly with true status", () => {
   const renderer = new ShallowRenderer();

   const task = {
      id: 3,
      status: true,
      text: "Catch Bulbasaur of type Grass,poison",
      imageURL: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
   };

   const tree = renderer.render(<Task task={task} />);
   expect(tree).toMatchSnapshot();
});

it("renders pokemon task correctly with false status", () => {
   const renderer = new ShallowRenderer();

   const task = {
      id: 3,
      status: false,
      text: "Catch Bulbasaur of type Grass,poison",
      imageURL: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
   };

   const tree = renderer.render(<Task task={task} />);
   expect(tree).toMatchSnapshot();
});

it("renders regular task correctly without status", () => {
   const renderer = new ShallowRenderer();

   const task = {
      id: 3,
      text: "Hello",
   };

   const tree = renderer.render(<Task task={task} />);
   expect(tree).toMatchSnapshot();
});

it("renders regular task correctly with status true", () => {
   const renderer = new ShallowRenderer();

   const task = {
      id: 3,
      text: "Hello",
      status: true,
   };

   const tree = renderer.render(<Task task={task} />);
   expect(tree).toMatchSnapshot();
});
