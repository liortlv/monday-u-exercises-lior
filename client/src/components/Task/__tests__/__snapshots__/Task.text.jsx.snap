// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders pokemon task correctly with false status 1`] = `
<ForwardRef(MotionComponent)
  animate={
    Object {
      "opacity": 1,
      "scale": 1,
      "x": 0,
    }
  }
  className="task-container"
  data-id={3}
  initial={
    Object {
      "opacity": 0.3,
      "scale": 0.6,
      "x": -130,
    }
  }
  whileHover={
    Object {
      "scale": 1.05,
      "transition": Object {
        "duration": 0.15,
      },
    }
  }
>
  <Memo(Connect(Checkbox))
    task={
      Object {
        "id": 3,
        "imageURL": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
        "status": false,
        "text": "Catch Bulbasaur of type Grass,poison",
      }
    }
  />
  <div
    className="task"
  >
    Catch Bulbasaur of type Grass,poison
    <img
      alt="pokemon"
      className="pokemon-image"
      src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
    />
  </div>
  <Button
    className="delete"
    innerText={
      <i
        className="fa-trash fa-trash fa-solid"
      />
    }
    onClick={[Function]}
  />
</ForwardRef(MotionComponent)>
`;

exports[`renders pokemon task correctly with true status 1`] = `
<ForwardRef(MotionComponent)
  animate={
    Object {
      "opacity": 1,
      "scale": 1,
      "x": 0,
    }
  }
  className="task-container"
  data-id={3}
  initial={
    Object {
      "opacity": 0.3,
      "scale": 0.6,
      "x": -130,
    }
  }
  whileHover={
    Object {
      "scale": 1.05,
      "transition": Object {
        "duration": 0.15,
      },
    }
  }
>
  <Memo(Connect(Checkbox))
    task={
      Object {
        "id": 3,
        "imageURL": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
        "status": true,
        "text": "Catch Bulbasaur of type Grass,poison",
      }
    }
  />
  <div
    className="task"
  >
    Catch Bulbasaur of type Grass,poison
    <img
      alt="pokemon"
      className="pokemon-image"
      src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
    />
  </div>
  <Button
    className="delete"
    innerText={
      <i
        className="fa-trash fa-trash fa-solid"
      />
    }
    onClick={[Function]}
  />
</ForwardRef(MotionComponent)>
`;

exports[`renders regular task correctly with status true 1`] = `
<ForwardRef(MotionComponent)
  animate={
    Object {
      "opacity": 1,
      "scale": 1,
      "x": 0,
    }
  }
  className="task-container"
  data-id={3}
  initial={
    Object {
      "opacity": 0.3,
      "scale": 0.6,
      "x": -130,
    }
  }
  whileHover={
    Object {
      "scale": 1.05,
      "transition": Object {
        "duration": 0.15,
      },
    }
  }
>
  <Memo(Connect(Checkbox))
    task={
      Object {
        "id": 3,
        "status": true,
        "text": "Hello",
      }
    }
  />
  <div
    className="task"
  >
    Hello
    
  </div>
  <Button
    className="delete"
    innerText={
      <i
        className="fa-trash fa-trash fa-solid"
      />
    }
    onClick={[Function]}
  />
</ForwardRef(MotionComponent)>
`;

exports[`renders regular task correctly without status 1`] = `
<ForwardRef(MotionComponent)
  animate={
    Object {
      "opacity": 1,
      "scale": 1,
      "x": 0,
    }
  }
  className="task-container"
  data-id={3}
  initial={
    Object {
      "opacity": 0.3,
      "scale": 0.6,
      "x": -130,
    }
  }
  whileHover={
    Object {
      "scale": 1.05,
      "transition": Object {
        "duration": 0.15,
      },
    }
  }
>
  <Memo(Connect(Checkbox))
    task={
      Object {
        "id": 3,
        "text": "Hello",
      }
    }
  />
  <div
    className="task"
  >
    Hello
    
  </div>
  <Button
    className="delete"
    innerText={
      <i
        className="fa-trash fa-trash fa-solid"
      />
    }
    onClick={[Function]}
  />
</ForwardRef(MotionComponent)>
`;
